cloud/docker/docker_swarm_service.py

```diff
@@ -0,0 +1,795 @@
+      'endpoint_mode'           : self.endpoint_mode,
+      'restart_policy'          : self.restart_policy,
+      'limit_cpu'               : self.limit_cpu,
+      'limit_memory'            : self.limit_memory,
+      'reserve_cpu'             : self.reserve_cpu,
+      'reserve_memory'          : self.reserve_memory,
+      'restart_policy_delay'    : self.restart_policy_delay,
+      'restart_policy_attempts' : self.restart_policy_attempts,
+      'restart_policy_window'   : self.restart_policy_window
+    }
+
+  @staticmethod
+  def from_ansible_params(ap):
+    s=DockerService()
+    s.constraints              = ap['constraints']
+    s.image                    = ap['image']
+    s.args                     = ap['args']
+    s.endpoint_mode            = ap['endpoint_mode']
+    s.env                      = ap['env']
+    s.labels                   = ap['labels']
+    s.container_labels         = ap['container_labels']
+    s.limit_cpu                = ap['limit_cpu']
+    s.limit_memory             = ap['limit_memory']
+    s.reserve_cpu              = ap['reserve_cpu']
+    s.reserve_memory           = ap['reserve_memory']
+    s.mode                     = ap['mode']
+    s.mounts                   = ap['mounts']
+    s.constraints              = ap['constraints']
+    s.networks                 = ap['networks']
+    s.publish                  = ap['publish']
+    s.replicas                 = ap['replicas']
+    s.restart_policy           = ap['restart_policy']
+    s.restart_policy_attempts  = ap['restart_policy_attempts']
+    s.restart_policy_delay     = ap['restart_policy_delay']
+    s.restart_policy_window    = ap['restart_policy_window']
+    s.user                     = ap['user']
+
+    for param_name in ['reverve_memory', 'reserve_cpu',
+                       'limit_memory'  , 'limit_cpu']:
+      if ap.get(param_name):
+        try:
+          setattr(s, param_name, human_to_bytes(ap[param_name]))
+        except ValueError as exc:
+          raise Exception("Failed to convert %s to bytes: %s" % (param_name, exc))
+
+    s.publish=[]
+    for param_p in ap['publish']:
+      service_p = {}
+      service_p['protocol']       = param_p.get( 'protocol', 'tcp' )
+      service_p['published_port'] = param_p['published_port']
+      service_p['target_port']    = param_p['target_port']
+      if service_p['protocol'] not in ['tcp', 'udp']:
+        raise ValueError("got publish.protocol '%s', valid values:'tcp', 'udp'" %
+                         service_p['protocol'])
+      s.publish.append(service_p)
+    s.mounts=[]
+    for param_m in ap['mounts']:
+      service_m = {}
+      service_m['readonly'] = bool(param_m.get('readonly', False))
+      service_m['type']     = param_m.get('type', 'bind')
+      service_m['source']   = param_m['source']
+      service_m['target']   = param_m['target']
+      s.mounts.append(service_m)
+    return s
+
+  def compare(self,os):
+    differences = []
+    can_edit_live = True
```

Comment:
The ```needs_rebuild``` logic does not work. If the service exists it is always rebuilt, because this value is always ```True```. Changing this line to ```can_edit_live = False``` fixes it.